src/loader.cpp

```diff
@@ -179,16 +180,44 @@ void loader_local::initialize(const json& config_json)
 
     m_provider = provider_factory::create(config_json);
 
-    unsigned int threads_num = lcfg.decode_thread_count != 0 ? lcfg.decode_thread_count
-                                                             : std::thread::hardware_concurrency();
+    // Parse list of cpu cores to set thread affinity.
+    const char*      env_cpu_list = getenv("AEON_CPU_LIST"); // Example: 0-4,30,10,32-33
+    std::vector<int> thread_affinity_map;
+    if (env_cpu_list != nullptr && std::strlen(env_cpu_list) > 0) // set affinity from environment
+    {
+        std::string str_cpu_list{env_cpu_list};
+        thread_affinity_map = nervana::parse_cpu_list(str_cpu_list);
+
+        if (!lcfg.cpu_list.empty())
+        {
+            WARN << "Both AEON_CPU_LIST and 'cpu_list' are detected. AEON_CPU_LIST environment "
+                    "variable will be used in this case.";
+        }
+    }
+    else if (!lcfg.cpu_list.empty()) // set affinity from config
+    {
+        thread_affinity_map = nervana::parse_cpu_list(lcfg.cpu_list);
+    }
+    if (thread_affinity_map.empty()) // automatically determine number of threads
+    {
+        // we don't use all threads, some of them we leave for other pipeline objects and system
+        auto nthreads =
+            std::thread::hardware_concurrency() -
+            std::min(m_max_count_of_free_threads,
+                     static_cast<int>(std::thread::hardware_concurrency() / m_free_threads_ratio));
+        thread_affinity_map.resize(nthreads);
+        std::iota(thread_affinity_map.begin(), thread_affinity_map.end(), 0);
+    }
+    unsigned int threads_num = !thread_affinity_map.empty() ? thread_affinity_map.size()
```

Comment:
I think here you don't need to check for _non emptyness_ of `thread_affinity_map`. In above code you fill it in case there is no `cpu_list` nor `AEON_CPU_LIST` variables set. Thus I think the following line would suffice:
```suggestion
    unsigned int threads_num = thread_affinity_map.size();
```
Moreover If `thread_affinity_map.empty()` would be true, why setting `threads_num` to `std::thread::hardware_concurrency()` ? I wonder whether this would be correct

Reply:
Thanks @arogowie-intel . You are right. There is something wrong. I will look closer at this because I'm not sure what @AdrShv meant here for threads_num variable used below in line 215 here https://github.com/NervanaSystems/aeon/pull/9/files/b70b736b96c97fc5b0f152bc06667dbf8475d99c#diff-50f539ac61d3b1b208f9d379fdea5153R215. It has different calculation than https://github.com/NervanaSystems/aeon/pull/9/files/b70b736b96c97fc5b0f152bc06667dbf8475d99c#diff-50f539ac61d3b1b208f9d379fdea5153R204 so I think it should still be a different number but I got it wrong.